lang/py/src/avro/codecs.py

```diff
@@ -0,0 +1,181 @@
+#!/usr/bin/env python
+
+##
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+# https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import, division, print_function
+
+import struct
+import sys
+import zlib
+from binascii import crc32
+
+from avro import io, schema
+
+#
+# Constants
+#
+if sys.version_info >= (2, 5, 0):
+  struct_class = struct.Struct
+else:
+  class SimpleStruct(object):
+    def __init__(self, format):
+      self.format = format
+    def pack(self, *args):
+      return struct.pack(self.format, *args)
+    def unpack(self, *args):
+      return struct.unpack(self.format, *args)
+  struct_class = SimpleStruct
+
+STRUCT_CRC32 = struct_class('>I')  # big-endian unsigned int
+
+
+try:
+  from cStringIO import StringIO
+except ImportError:
+  from StringIO import StringIO
+try:
+  import bz2
+  has_bzip2 = True
+except ImportError:
+  has_bzip2 = False
+try:
+  import snappy
+  has_snappy = True
+except ImportError:
+  has_snappy = False
+try:
+  import zstandard as zstd
+  has_zstandard = True
+except ImportError:
+  has_zstandard = False
+
+
+class Codec(object):
+  def compress(self, data):
+    raise NotImplementedError()
```

Comment:
This is fine, but fyi `raise` will instantiate the exception for you, so you don't need the parentheses.

```
raise NotImplementedError
```

is equivalent, and `raise Exception, "oops"` is the same as `raise Exception("oops")`, but has very slightly better parsing performance.

Reply:
Also we shouldn't skimp on the docstrings if we want others to feel comfortable contributing future compression codecs.